p2p/src/net/mock/floodsub.rs

```diff
@@ -0,0 +1,549 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): A. Altonen
+#![allow(unused)]
+#![allow(clippy::type_complexity)]
+
+use crate::{
+    error::{self, P2pError},
+    message::Message,
+    net::mock::{floodsub, types},
+    net::{mock::MockSocket, FloodsubTopic, NetworkService, SocketService},
+    peer::Peer,
+};
+use async_trait::async_trait;
+use futures::FutureExt;
+use logging::log;
+use parity_scale_codec::{Decode, Encode};
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap, HashSet},
+    hash::{Hash, Hasher},
+    io::{Error, ErrorKind},
+    net::{IpAddr, Ipv4Addr, SocketAddr},
+};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpListener, TcpStream},
+    sync::{broadcast, mpsc},
+};
+
+#[derive(Debug, PartialEq, Eq, Encode, Decode, Hash)]
+pub struct FloodsubMessage {
+    topic: FloodsubTopic,
+    message: Vec<u8>,
+}
+
+pub struct Floodsub {
+    /// TX channel for sending events to the frontend
+    event_tx: mpsc::Sender<types::FloodsubEvent>,
+
+    /// RX channel for receiving commands from the backend
+    cmd_rx: mpsc::Receiver<types::FloodsubCommand>,
+
+    /// Set of peers
+    peers: HashMap<SocketAddr, mpsc::Sender<()>>,
+
+    /// Set of subscriptions
+    subscriptions: HashMap<FloodsubTopic, HashSet<SocketAddr>>,
+
+    /// Broadcast channel for sending messages to the sockets
+    broadcast: (
+        broadcast::Sender<(FloodsubTopic, Vec<u8>)>,
+        broadcast::Receiver<(FloodsubTopic, Vec<u8>)>,
+    ),
+
+    /// Channels for receiving messages from the sockets
+    msg_chan: (
+        mpsc::Sender<(SocketAddr, FloodsubMessage)>,
+        mpsc::Receiver<(SocketAddr, FloodsubMessage)>,
+    ),
+
+    /// All messages that have been received and forwarded to frontend
+    seen_messages: HashSet<u64>,
```

Comment:
This is not necessary. In general there are much easier ways to ddos a node than sending the same message again and again.

Reply:
This is actually not for DDoS prevention but just that I don't spam the front-end with messages that it has already seen because for the "happy path" it has no use for them. But thinking about it again made me realize we need to forward also the duplicate messages so we can implement proper peer reputation.

As an aside though, this is the mock interface so IMO there could be a WinNuke here for all I care and fixing it wouldn't have a high priority unless it made the upper-level protocol implementation incorrect. But yeah, I should strive for correctness either way.